### PR TITLE
Part 2

### PR DESCRIPTION
**Pull request by Heartzian**

### CODE REVIEW

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Your code produces a lot of warnings when it is compiled and executed.
Please fix the warnings and re-request a review.

**Review by Heartzian — COMMENTED** [review]
Checked!

File: main.cpp [hunk 1]
```diff
@@ -125,84 +175,110 @@ struct CarWash
...
+    double mbProdsPrice = 1.25;
+    double lProdsPrice = 0.85;
+    double cbProdsPrice = 1.25;
+    double dProdsPrice = 0.9;
+    double memberDiscount = 0.9;
+    productsPrice = (breakfastProds * bfProdsPrice) +\
+                    (morningBreakProds * mbProdsPrice) +\
+                    (lunchProds * lProdsPrice) +\
+                    (coffeeBreakProds * cbProdsPrice) +\
+                    (dinnerProds * dProdsPrice);
```
Comment (matkatmusic) [inline main.cpp]: remove the `\`.  It is not necessary and just adds visual confusion to the code.
Semicolons indicate the end of a statement in C++, not newlines.
- [x] revise

File: main.cpp [hunk 2]
```diff
@@ -125,84 +175,110 @@ struct CarWash
...
+    }
+    totalToPay = (productsPrice * tax) + deliveryCharge;
+}
+
+void Market::sellProducts() //Nick example
+{ 
+    std::cout << "Dear Customer, you ordered " <<  nick.productsToOrder << "products, with a total cost of $" << nick.totalToPay << "which will be Delivered by Tom. Thanks for buying with us!";
+
+    numLogisticVehicles--;
+    numPeopleWorkingAtStore--;
```
Comment (matkatmusic) [inline main.cpp]: use pre-increment/decrement, not post-increment/decrement.
- [x] revise

File: main.cpp [hunk 3]
```diff
@@ -228,346 +305,521 @@ struct University
         std::string hobby = "Play Bass";
         int semestralCredits = 9;
 
-        float weekstudyTime(int research = 28, int teaching = 16);
-        float computeMonthlyExpenses(float food = 250, 
+        float weekStudyTime(int research = 28, int teaching = 16);
```
Comment (matkatmusic) [inline main.cpp]: Add a verb to this function name.  `computeWeekStudyTime` is a good choice
- [x] revise

File: main.cpp [hunk 4]
```diff
@@ -228,346 +305,521 @@ struct University
...
+    return canTrainAI = true;
+}
+
+bool Computer::Hardware::workAtOffice(std::string mostlyUsedTask)
+{
+    if (canPlayGames == true)
+    {
+        std::cout << "You can use this computer to do " << mostlyUsedTask;
+    }
+    return canUseAtOffice = true;
```
Comment (matkatmusic) [inline main.cpp]: Do not write code like this.
What value are you returning?
it is impossible to reason which value is being returned.
Separate the assignment `= true` from the value being returned.
```
canUseAtOffice = true;
return .......
```
- [x] revise

File: main.cpp [hunk 5]
```diff
@@ -228,346 +305,521 @@ struct University
...
+    }
+}
+
+bool Computer::saveInfo(bool diskAvailable)
+{
+    if (diskAvailable == true)
+    {
+        std::cout << "The file will be saved";
+    }
+    return canSaveInfo = true;
```
Comment (matkatmusic) [inline main.cpp]: Do not write code like this.
What value are you returning?
it is impossible to reason which value is being returned.
Separate the assignment `= true` from the value being returned.
```
canSaveInfo = true;
return .......
```
- [x] revise

File: main.cpp [hunk 6]
```diff
@@ -228,346 +305,521 @@ struct University
...
- 
+//Computer OfficeDesktop;
+
+bool Computer::Hardware::playGames(bool specs)
+{
+    if (specs == true && RAM >= 4 && disk >=256 && CPU >= 3.0 && GPU >= 2)
+    {
+        std::cout << "You can use this computer to play games";  
+    }
+    return canPlayGames = true;
```
Comment (matkatmusic) [inline main.cpp]: Do not write code like this.
What value are you returning?
it is impossible to reason which value is being returned.
Separate the assignment `= true` from the value being returned.
```
canPlayGames = true;
return .......
```
- [x] revise

File: main.cpp [hunk 7]
```diff
@@ -228,346 +305,521 @@ struct University
...
+}
+
+bool Computer::Hardware::trainAI(std::string softwareReq)
+{
+    if (canPlayGames == true && GPU > 4)
+    {
+        std::cout << "You can use this computer to train AI using" << softwareReq;
+    }
+    return canTrainAI = true;
```
Comment (matkatmusic) [inline main.cpp]: Do not write code like this.
What value are you returning?
it is impossible to reason which value is being returned.
Separate the assignment `= true` from the value being returned.
```
canTrainAI = true;
return .......
```
- [x] revise

File: main.cpp [hunk 8]
```diff
@@ -228,346 +305,521 @@ struct University
...
+    return food + fun + other;
+}
+
+float University::teachStudents(Professor) 
+{
+    int calculus = 8;
+    int programming = 12;
+    int cad = 6;
+    float teachedClasses = calculus + programming + cad;
+    return teachedClasses;
```
Comment (matkatmusic) [inline main.cpp]: Skip the temporary variable.
`return calculus + programming + cad;`
- [x] revise

File: main.cpp [hunk 9]
```diff
@@ -228,346 +305,521 @@ struct University
...
+}
+
+float University::Student::computeMonthlyExpenses(float food = 350,
+                                                  float fun, 
+                                                  float other)
+{
+    return food + fun + other;
+}
+
+float University::teachStudents(Professor)
```
Comment (matkatmusic) [inline main.cpp]: What is the point of declaring a function parameter if you're not going to use it?
- [x] revise: either use the function parameter or remove the function parameter

File: main.cpp [hunk 10]
```diff
@@ -228,346 +305,521 @@ struct University
...
-    2) overDrive the signal
-    3) give a better sonority to an instrument 
- */
+double Oscillator::generateSignal(Oscillator Square, double x)
+{
+    if (Square.waveformShape == "Square")
+    {
+        x = 0.5 * (frequency); //This is a dummy example
+    }
+    return x;
```
Comment (matkatmusic) [inline main.cpp]: Don't pass in an instance of the `Oscillator` class into an Oscillator member function!
If you're trying to use `waveformShape`, which is a member variable of the `Oscillator` instance being used when this function is called, just use it directly inside the function.   
```
if( waveformShape == "Square" )
    x = 0.5 * frequency;

return x;
```
- [x] revise

File: main.cpp [hunk 11]
```diff
@@ -228,346 +305,521 @@ struct University
...
+    return storedROM;
+}
+
+void Oscillator::playbackROMSamples(double x, bool anyKeyPressed)
+{
+    do 
+    {
+        x = storedROM + 1;
+    }
+    while (anyKeyPressed == true);
```
Comment (matkatmusic) [inline main.cpp]: Loops have not been covered yet. 
write something else
This function implementation easy creates an infinite loop if you call this function with `anyKeyPressed` set to `true`.
- [x] revise

File: main.cpp [hunk 12]
```diff
@@ -228,346 +305,521 @@ struct University
...
+}
 
 struct Filters 
 {
```
Comment (matkatmusic) [inline main.cpp]: This class represents ONE filter, not multiple filters
Name it that way
- [x] revise

File: main.cpp [hunk 13]
```diff
@@ -228,346 +305,521 @@ struct University
...
-    4) pre-delay (float)
-    5) size (float)
-3 things it can do:
-    1) simulate an space to a soun
-    2) give a better sound to an instrument
-    3) give depth to a band
- */
+double Filters::boostCutFreq(Filters LowShelf, double x)
+{
+    LowShelf.gain=-12;
```
Comment (matkatmusic) [inline main.cpp]: Do not pass in an instance of `Filter` into a `Filter` member function
You can use the member variables of the calling instance directly inside the member function. 
- [x] revise

File: main.cpp [hunk 14]
```diff
@@ -228,346 +305,521 @@ struct University
...
-    5) reverb
-3 things it can do:
-    1) generate a signal
-    2) modify the generated signal
-    3) process the modified signal
- */
+double Reverb::simulateSpace(Reverb Plate, double x)
+{
+    return x = x * (Plate.time + Plate.mix);
+}
```
Comment (matkatmusic) [inline main.cpp]: Do not pass a `Reverb` into a `Reverb` member function
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**